dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -0,0 +1,580 @@
+    for (;;)
+    {
+        advancePartitionEnd();
+
+        // Either we ran out of data or we found the end of partition (maybe
+        // both, but this only happens at the total end of data).
+        assert(partition_ended || partition_end == blocksEnd());
+        if (partition_ended && partition_end == blocksEnd())
+        {
+            assert(input_is_finished);
+        }
+
+
+        while (current_row < partition_end)
+        {
+            // if window only have row_number function, we can ignore judging peers
+            if (!only_have_row_number)
+            {
+                // We now know that the current row is valid, so we can update the
+                // peer group start.
+                if (!arePeers(peer_group_start, current_row))
+                {
+                    peer_group_start = current_row;
+                    peer_group_start_row_number = current_row_number;
+                    ++peer_group_number;
+                }
+            }
+
+            // Advance the frame start.
+            advanceFrameStart();
+
+            if (!frame_started)
+            {
+                // Wait for more input data to find the start of frame.
+                assert(!input_is_finished);
+                assert(!partition_ended);
+                return;
+            }
+
+            // Advance the frame end.
+            advanceFrameEnd();
+
+            if (!frame_ended)
+            {
+                // Wait for more input data to find the end of frame.
+                assert(!input_is_finished);
+                assert(!partition_ended);
+                return;
+            }
+
+            // The frame can be empty sometimes, e.g. the boundaries coincide
+            // or the start is after the partition end. But hopefully start is
+            // not after end.
+            assert(frame_started);
+            assert(frame_ended);
+            assert(frame_start <= frame_end);
+
+            prev_frame_start = frame_start;
+
+            // Write out the results.
+            writeOutCurrentRow();
+
+            if (isCancelledOrThrowIfKilled())
+            {
+                if (!window_blocks.empty())
+                    window_blocks.erase(window_blocks.begin(), window_blocks.end());
+                input_is_finished = true;
+                return;
+            }
+
+            // Move to the next row. The frame will have to be recalculated.
+            // The peer group start is updated at the beginning of the loop,
+            // because current_row might now be past-the-end.
+            advanceRowNumber(current_row);
+            ++current_row_number;
+            first_not_ready_row = current_row;
+            frame_ended = false;
+            frame_started = false;
+        }
+
+        if (input_is_finished)
+        {
+            // We finalized the last partition in the above loop, and don't have
+            // to do anything else.
```

Comment:
Does this mean all the job is done? If yes, add an assert that `current_row == blocksEnd()`?

Reply:
> Does this mean all the job is done? If yes, add an assert that `current_row == blocksEnd()`?
done